### PR TITLE
[Haxe 3.3] Adds Support of -D display-stdin. closes #1311

### LINKED ISSUE #1311: [Haxe] Support of -D display-stdin
https://github.com/HaxeFoundation/haxe/pull/5120

### PR DESCRIPTION
**Pull request by SlavaRa**

### CODE REVIEW

**Comment by Neverbirth** [comment]
Hmmm, what about multiple unsaved files at the same time? It seems it only gets the content of the current file are the rest are ignored. Am I right? Can we pass the contents of multiple files or only the one you want to get completion from? if it's the second then this seems to have severe drawbacks, doesn't it?

**Comment by nadako** [comment]
It's the second. We thought about this, but decided not to overcomplicate things for now, since the main issue was that IDEs basically had to save the current file on each keystroke and this fixes it.

To do it "properly", we would have to do something like e.g. Roslyn does, when source files are synchronized in the compiler with IDE's in-memory contents using some protocol. Haxe is not ready for this right now, so the current display-stdin solution is the best we can do for the time being.

**Comment by nadako** [comment]
By the way, I'm not really familiar with FD's codebase, but it looks like this PR doesn't implement passing stdin in the non-server mode, which can also be done (you just call `haxe -D display-stdin --display ...` and write contents to the standard input stream of the process.

**Comment by nadako** [comment]
(see description here https://github.com/HaxeFoundation/haxe/pull/5120)

**Comment by Neverbirth** [comment]
Then I guess FD should still save the other files to be on the safe side?

**Comment by SlavaRa** [comment]
I think that I could save other modified files, but not current file. @Neverbirth  what you this about this?

**Comment by Neverbirth** [comment]
It seems we wrote mostly at the same time, yeah, I'm OK with that one.

**Comment by SlavaRa** [comment]
@Neverbirth fixed

**Comment by elsassph** [comment]
When multiple files are modified a nice solution could be to Save all modified excepted the current one.

**Comment by elsassph** [comment]
Amazing stuff guys

**Comment by SlavaRa** [comment]
@elsassph @Neverbirth @Meychi what about this PR?

File: External/Plugins/HaXeContext/Context.cs [hunk 1]
```diff
@@ -1209,6 +1209,14 @@ public override MemberList ResolveDotContext(ScintillaNet.ScintillaControl sci,
             return null; 
         }
 
+        HaxeComplete GetHaxeComplete(ScintillaControl sci, ASExpr expression, bool autoHide, HaxeCompilerService compilerService)
+        {
+            var sdkVersion = GetCurrentSDKVersion();
+            if (hxsettings.CompletionMode == HaxeCompletionModeEnum.CompletionServer && new SemVer("3.2.1").IsOlderThan(sdkVersion))
```
Comment (elsassph) [inline External/Plugins/HaXeContext/Context.cs]: Maybe do a semver greater-than-or-equal 3.3.0 if it's possible.

File: External/Plugins/HaXeContext/HaxeComplete330.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            var args = base.BuildHxmlArgs();
+            if (args == null) return null;
+            var list = new List<string>(args) {"-D display-stdin"};
+            var result = list.ToArray();
+            return result;
+        }
+
+        protected override string GetFileContent()
+        {
+            return Sci.Text;
```
Comment (elsassph) [inline External/Plugins/HaXeContext/HaxeComplete330.cs]: Is there any possible filesize concern here? What about if you're editing a very large file?

File: External/Plugins/HaXeContext/Context.cs (hunk 1 above)
Comment (SlavaRa) [inline External/Plugins/HaXeContext/Context.cs]: Fixed https://github.com/fdorg/flashdevelop/pull/1318/commits/7c68005059ebd11b13cd1a3f65420fd4ead20689

File: External/Plugins/HaXeContext/HaxeComplete330.cs (hunk 2 above)
Comment (SlavaRa) [inline External/Plugins/HaXeContext/HaxeComplete330.cs]: My test file contains more than 20,000 lines of text and all ok)

**Pull request merged**